File: packages/shared/src/badges/badge-kind.ts
/**
 * The inline badge kinds — the app's signature interaction, enumerated.
 *
 * Purpose
 *   `docs/product/prd.md` "Tab 2" defines the small glowing pills that sit inline in
 *   scripture and open a focused sheet. Every one of them needs a stable discriminator
 *   the reader, the API, and the sheet router all agree on. This module is that list.
 *
 * Key responsibilities
 *   - Enumerate the badge kinds as a closed union, so a `switch` over them is exhaustive.
 *   - Carry each kind's product-level presentation: bracketed label, glyph, and what it
 *     appears beside. These come from the spec, not from the design system — the hue is
 *     a design token and deliberately lives in the theme layer, not here.
 *
 * Dependencies
 *   None. Pure data.
 *
 * Count discrepancy — recorded, not resolved (question `Q-018`)
 *   The PRD's prose says "10 Embedded Feature Badges" but the list beneath it contains
 *   **eleven** marks, because Route and 3D City share one bullet. Separately,
 *   `docs/product/design-language.md` §2 assigns a hue to ten badges and omits Lineage,
 *   even though the PRD gives Lineage its own bullet *and* its own sheet ("3D family
 *   tree node graph"). This module models all eleven: a badge type with a specified
 *   sheet but no assigned hue is a design gap to fill, whereas dropping it would be a
 *   feature silently lost.
 */

/**
 * Which inline badge this is. The discriminant of the `InlineBadge` union.
 *
 * Values are kebab-case and stable — they appear in pre-computed passage records and in
 * analytics events, so renaming one is a data migration, not a refactor.
 */
export type BadgeKind =
  | 'route'
  | '3d-city'
  | 'history'
  | 'root'
  | 'lineage'
  | 'manuscript'
  | 'cross-ref'
  | 'structure'
  | 'cultural'
  | 'context'
  | 'meditate';

/** The product-level description of one badge kind. */
export interface BadgeKindDescriptor {
  /** The discriminator, repeated here so an iterated descriptor is self-contained. */
  readonly kind: BadgeKind;
  /** The word inside the brackets, e.g. `Route` in `[🗺 Route]`. */
  readonly label: string;
  /** The emoji the spec pairs with the label. Part of the mark, per the design language. */
  readonly glyph: string;
  /** What the badge annotates, condensed from the PRD. Guides the enrichment pipeline. */
  readonly appearsBeside: string;
}

/** Every badge kind, in the order the PRD lists them. */
export const BADGE_KINDS = [
  'route',
  '3d-city',
  'history',
  'root',
  'lineage',
  'manuscript',
  'cross-ref',
  'structure',
  'cultural',
  'context',
  'meditate',
] as const satisfies readonly BadgeKind[];

/**
 * Each kind's presentation, keyed by the discriminator.
 *
 * A `Record` over the literal union rather than a `Map`: the compiler then refuses to
 * merge a new kind into `BadgeKind` without also describing it here, so the two can
 * never drift apart.
 */
const DESCRIPTOR_BY_KIND: Readonly<Record<BadgeKind, BadgeKindDescriptor>> = {
  route: { kind: 'route', label: 'Route', glyph: '🗺️', appearsBeside: 'location names' },
  // `Q-025`: the label is what the reader is shown, and the sheet behind it shows the
  // gazetteer site record rather than a reconstruction — no openly licensed 3D model of a
  // biblical city exists (`DECISIONS.md` §4). The discriminator keeps its published
  // spelling; only the words the reader reads changed.
  '3d-city': {
    kind: '3d-city',
    label: 'Site',
    glyph: '🏛️',
    appearsBeside: 'cities named in the passage',
  },
  history: {
    kind: 'history',
    label: 'History',
    glyph: '⏳',
    appearsBeside: 'historical rulers or laws',
  },
  root: { kind: 'root', label: 'Root', glyph: '🗣️', appearsBeside: 'key original-language words' },
  lineage: { kind: 'lineage', label: 'Lineage', glyph: '🧬', appearsBeside: 'biblical figures' },
  manuscript: {
    kind: 'manuscript',
    label: 'Manuscript',
    glyph: '📜',
    appearsBeside: 'famous textual variants',
  },
  'cross-ref': {
    kind: 'cross-ref',
    label: 'Cross-Ref',
    glyph: '🎯',
    appearsBeside: 'quotations and fulfilled prophecies',
  },
  structure: {
    kind: 'structure',
    label: 'Structure',
    glyph: '🌳',
    appearsBeside: 'chiasms, Psalms and Proverbs',
  },
  cultural: {
    kind: 'cultural',
    label: 'Cultural',
    glyph: '⚖️',
    appearsBeside: 'ancient Near East and Greco-Roman customs',
  },
  context: {
    kind: 'context',
    label: 'Context',
    glyph: '🎙️',
    appearsBeside: 'verses needing historical background',
  },
  meditate: {
    kind: 'meditate',
    label: 'Meditate',
    glyph: '🧘',
    appearsBeside: 'high-impact devotional or command verses',
  },
};

/**
 * Every descriptor, in PRD order.
 *
 * Iterate this rather than hand-writing a list; the chapter-end badge summary and the
 * enrichment pipeline's coverage report both depend on it being complete.
 */
export const BADGE_KIND_DESCRIPTORS: readonly BadgeKindDescriptor[] = BADGE_KINDS.map(
  (kind) => DESCRIPTOR_BY_KIND[kind],
);

/** How many inline badge kinds exist. See the count discrepancy note above. */
export const BADGE_KIND_COUNT = BADGE_KINDS.length;

/**
 * Narrow an arbitrary string to a badge kind.
 *
 * Pre-computed passage records are authored outside the app, so a `kind` arriving from
 * one is untrusted input and must be checked at the boundary (rule 6.5.1).
 *
 * @param value - A candidate discriminator, typically from JSON.
 * @returns `true` when the value is one of the known kinds. Side effects: none.
 */
export function isBadgeKind(value: string): value is BadgeKind {
  return BADGE_KINDS.some((kind) => kind === value);
}

/**
 * Look up a badge kind's label, glyph, and placement rule.
 *
 * @param kind - A known badge kind.
 * @returns Its descriptor. Total over `BadgeKind`, so there is no failure branch.
 *
 * Side effects: none.
 */
export function describeBadgeKind(kind: BadgeKind): BadgeKindDescriptor {
  return DESCRIPTOR_BY_KIND[kind];
}

File: apps/api/app/modules/badges/domain/badge_kind.py
"""The five badge kinds M2 ships, and the order they win arguments in.

Purpose
    `packages/shared/src/badges/badge-kind.ts` enumerates eleven kinds for the
    whole product. Decision P-04 selects five of them for this milestone, and
    only those five have deterministic data behind them today. This module is
    that subset, plus the total order two colliding badges are resolved by.

Key responsibilities
    - Name the five kinds with the exact wire strings the TypeScript union uses.
      These are the discriminant the client switches on; renaming one is a
      breaking change, not a refactor.
    - Define badge priority, so a collision has one answer and not two.

Dependencies
    Standard library only. This is the innermost module of the badge domain.

Usage
    BadgeKind.ROUTE.value        # "route"
    priority_of(BadgeKind.ROOT)  # 3
"""

from __future__ import annotations

from enum import StrEnum


class BadgeKind(StrEnum):
    """A badge kind shipping in M2.

    Values match `BadgeKind` in `packages/shared/src/badges/badge-kind.ts`
    character for character. A test asserts that, so the two cannot drift.
    """

    ROUTE = "route"
    CITY_3D = "3d-city"
    HISTORY = "history"
    ROOT = "root"
    CROSS_REF = "cross-ref"


#: The five kinds in decision P-04's own listing order.
#:
#: That order is reused as the priority order, which is not a coincidence
#: dressed up as a rule: P-04 lists the spatial badges first because they are
#: the ones anchored to a proper noun the reader can see on the map, and the
#: verse-level badges last because they annotate the whole verse rather than a
#: word. When two badges want the same run of characters, the one making the
#: more specific claim about those characters should win.
M2_BADGE_KINDS: tuple[BadgeKind, ...] = (
    BadgeKind.ROUTE,
    BadgeKind.CITY_3D,
    BadgeKind.HISTORY,
    BadgeKind.ROOT,
    BadgeKind.CROSS_REF,
)

_PRIORITY: dict[BadgeKind, int] = {kind: index for index, kind in enumerate(M2_BADGE_KINDS)}


def priority_of(kind: BadgeKind) -> int:
    """Rank a kind for collision resolution. Lower wins.

    @param kind: Any M2 badge kind.
    @returns Its 0-based position in `M2_BADGE_KINDS`. Total over the enum, so
        there is no failure branch. Side effects: none.
    """
    return _PRIORITY[kind]


def parse_badge_kind(value: str) -> BadgeKind | None:
    """Narrow an untrusted string to a badge kind.

    Badge ids arrive from the client in a URL path, so the kind embedded in one
    is untrusted input and is checked at the boundary (rule 6.5.1).

    @param value: A candidate discriminant.
    @returns The kind, or None when the string names no M2 kind.
        Side effects: none.
    """
    for kind in M2_BADGE_KINDS:
        if kind.value == value:
            return kind
    return None
